### ISSUE TITLE
log4j and galaxy tool dependencies

### ISSUE THREAD

**Issue by colindaven** [galaxyproject/galaxy#13100]
Dear Galaxy team,

we have a local Galaxy in our internal network and have scanned for log4j vulnerabilities using log4jdetector https://github.com/mergebase/log4j-detector. 

There are quite a few java tools there - see below. 

What do you think we should do ? Remove these packages? Replace with updates where available?

Because Galaxy is a web server it's more likely to be a target in my understanding, correct ? Yet bioinformatics apps aren't too well known so are likely to be less targeted.

Thanks for any guidance,
Colin

```
tool_dependencies
{"hits":[
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__emboss@5.0.0/share/EMBOSS/jemboss/lib/axis/log4j-1.2.8.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__emboss@6.6.0/share/EMBOSS/jemboss/lib/axis/log4j-1.2.8.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__multigps@0.74/share/multigps-0.74-3/multigps_v0.74.jar","info":"contains Log4J-2.x   >= 2.0-beta9 (< 2.10.0)"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__mztosqlite@2.0.4/share/mztosqlite-2.0.4-3/mzToSQLite-2.0.4.jar","info":"contains Log4J-2.x   >= 2.10.0"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__peptide-shaker@1.16.36/share/peptide-shaker-1.16.36-0/lib/log4j-1.2.12.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__picard@2.18.2/share/picard-2.18.2-0/picard.jar","info":"contains Log4J-2.x   >= 2.0-beta9 (< 2.10.0)"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__qualimap@2.2.2c/share/qualimap-2.2.2c-1/paxtools-1.0b1.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__qualimap@2.2.2d/share/qualimap-2.2.2d-2/paxtools-1.0b1.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__snpeff@4.3.1t/share/snpeff-4.3.1t-0/snpEff.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__snpeff@4.5covid19/share/snpeff-4.5covid19-2/snpEff.jar","info":"contains Log4J-1.x AND Log4J-2.x _CRAZY_   >= 2.0-beta9 (< 2.10.0)"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/__tb-profiler@2.8.14/share/gatk4-4.2.0.0-1/gatk-package-4.2.0.0-local.jar","info":"contains Log4J-1.x AND Log4J-2.x _CRAZY_   >= 2.0-beta9 (< 2.10.0)"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-29b7bf6084468405ad5efc371adbfa9588dd7492a92e701dc78e5793340f4a89/share/picard-2.18.2-0/picard.jar","info":"contains Log4J-2.x   >= 2.0-beta9 (< 2.10.0)"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-2d71e7294258e0d6557ab09c36c8336300de41c90ae1e21e30396938913db7d4/share/gatk4-4.1.7.0-0/gatk-package-4.1.7.0-local.jar","info":"contains Log4J-1.x AND Log4J-2.x _CRAZY_   >= 2.0-beta9 (< 2.10.0)"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-5b8d730bcd7cf4907b19fb0d5ae836918bf9c31e160f369823d77f1272ed5d37/binaries/noarch/snpEff.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-5b8d730bcd7cf4907b19fb0d5ae836918bf9c31e160f369823d77f1272ed5d37/share/snpeff-5.0-1/snpEff.jar","info":"contains Log4J-1.x AND Log4J-2.x _CRAZY_   >= 2.0-beta9 (< 2.10.0)"},
{"_VULNERABLE_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-626e8c85236e11587c90fe300d7273816bacf20561a301d26c9ed80103b79b1a/share/picard-2.18.2-0/picard.jar","info":"contains Log4J-2.x   >= 2.0-beta9 (< 2.10.0)"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-6741b2041536887c38717699ddbd4400a50e29c203151eabe9a7f30600deaba3/share/EMBOSS/jemboss/lib/axis/log4j-1.2.8.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-98af0aba661e35478053dd4aed4f5b4b3c20f395d91e6fc0fa6ea55a0e8427f1/share/searchgui-3.3.10-0/lib/com.springsource.org.apache.log4j-1.2.15.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-98af0aba661e35478053dd4aed4f5b4b3c20f395d91e6fc0fa6ea55a0e8427f1/share/searchgui-3.3.10-0/lib/log4j-1.2.12.jar","info":"contains Log4J-1.x   <= 1.2.17"},
{"_OLD_":"/mnt/ngsnfs/rcug_galaxy/2021_vm11/tool_dependencies/_conda/envs/mulled-v1-b85d64533b49caecb64e870c19c004276c233b3f2091cdb79c95c7c4d19c6608/share/snpeff-4.3.1t-5/snpEff.jar","info":"contains Log4J-1.x   <= 1.2.17"},
```

**Comment by quacksawbones** [comment]
*\*DISCLAIMER\* - I'm not an InfoSec expert. Happy to be directed elsewhere/update my information if incorrect*

tl;dr: Galaxy itself isn't written in Java and so shouldn't be affected. Underlying tools that use Java might be.

There's some discussion about this in the Galaxy Admin Gitter - https://gitter.im/galaxyproject/admins?at=61c230e4e002586c753b0be6

My understanding of this is that any publicly-facing systems using modules/software with the log4j vulnerability are most at risk, as these are systems are where a bad actor may get in and start fishing for information. For an "internal" system (in our case, a Galaxy VM), said actor would first need to get VM information via a publicly-facing server, after which they can start weaseling their way inside. The way your publicly facing system is configured (and whether the log4j vulnerability has been adequately addressed or not) will determine your exposure risk more than Galaxy. (NB: This does not take into account the scenario of malware that is already inside and harvesting information).

But because Galaxy itself is written in Python, it shouldn't be the weak link. Regarding any bioinformatics tools themselves, they might be vulnerable if they aren't maintained (like the ones found in your list above). However, I suspect someone nefarious enough to exploit the log4j vulnerability in the first place would be smart enough to build something that can start sniffing around for any other vulnerable software and modules your system has available.

I'm not saying "It's not an issue", more that I believe:
- Galaxy itself should be fine,
- Any tools that Galaxy (or your Bioinformatics server) runs may be a risk if they're Java based and outdated,
- The risk diminishes:
   - The "further away" your internal systems are from a publicly facing one,
   - If your publicly-facing systems have been adequately patched, and
   - if a bad actor has already used this information to give themselves another way in.

Regardless, I am going to speak to our SysAdmin shortly and discuss this in relation to Galaxy and our organisation. I am happy to include any relevant information and/or suggestions they have here if that would be helpful (and doesn't inadvertently spam the Galaxy Devs)?

**Comment by colindaven** [comment]
Thanks @quacksawbones , I think this is a reasonable assessment of the situation. Of course, there can be a substantial gap between those working on the topic, and those (especially political level actors) managing responses.

**Comment by bgruening** [comment]
usegalaxy.eu also investigated this and we also concluded that we are not affected, similar to @quacksawbones writeup!
https://galaxyproject.eu/posts/2021/12/16/log4j/
